### ISSUE TITLE
auto_rate - "object of type 'closure' is not subsettable" error

### ISSUE THREAD

**Issue by nicholascarey** [januarharianto/respR#114]
This is a common enough error that i've seen it several times and still occurs in the latest version. I've tracked it down to 

`auto_rate_funs` > `auto_rate_linear` > `kernel_method` 

This line:
`peaks <- which(diff(sign(diff(d$y))) == -2) + 1`

Error occurs when this returns a `numeric(0)` instead of an index because two adjacent values where the peak is have identical values. 

See attached data. 

```
dt_err <- readRDS("dt_err.rda")  
width = 17   
top_only = TRUE
```

Then run through `kernel_method` code line by line. That line above returns `numeric(0)`.  Looking at `d$y` these values are identical:

```
> d$y[256:257]
[1] 35.07332 35.07332
```

so this line fails.... 

Not sure how to fix. Could be a simple but ugly solution. Like first thing, `sapply` through it and if any adjacent values are identical fuzz one up by a tiny value to make them not... 

Or probably a better way of finding a local peak. 

[This](https://github.com/nicholascarey/caRey#peaks) is probably overkill :)

But some of the code in it might help, e.g. :

```
  peaks <- zoo::rollapply(df, width = span * 2 + 1, function(q) {
    cent <- q[length(q[,1]) - span,1] # value to test
    x <- q[,2] ## all values
    # is it the highest value?
    # This also shoudl catch identical values, but only return first one as the peak
    logi <- all(x[cent] > x[1:(cent-1)]) && all(x[cent] >= x[(cent+1):length(x)])
    return(logi)
  },
  align = "center", ## so window is aligned on central value
  partial = TRUE, ## so it includes start/end where window is truncated
  by.column = FALSE) ## to apply across the df by row not column
```

Been too long since i wrote that function to be able to mentally parse it right now, but something in [there](https://github.com/nicholascarey/caRey/blob/master/R/peaks.R) might help.... 

[dt_err.rda.zip](https://github.com/januarharianto/respR/files/6552748/dt_err.rda.zip)

**Comment by nicholascarey** [comment]
Was going to make this a new issue, but i think it's related, because as far as i can tell it stems from the exact same line of code. 

```
auto_rate(zeb_intermittent.rd[20540:20960, ], plot = FALSE)

## fails with:

Error in set(x, j = name, value = value) : 
  Supplied 49 items to be assigned to 48 items of column 'density'. If you wish to 'recycle' the RHS please use rep() to make this intent clear to readers of your code. 
```

Comes from `kernel_method` producing 48 results (well, there are 48 `$subsets`), but this same line of code :

```
peaks <- which(diff(sign(diff(d$y))) == -2) + 1
```

identifies 49 peaks.  Error occurs in main `auto_rate` function when it attempts to add these 49 peaks to the 48 row `out$summary` element. 


This example comes straight from trying to run the code in [this vigette](https://januarharianto.github.io/respR/articles/intermittent2.html), so it must have been introduced in a (relatively) recent update.

**Comment by nicholascarey** [comment]
This code from the auto_rate performance vignette has this error for every attempt at `test_lin`. Current version in `develop` only, v1.1 works fine.... 

```{r benchmarks, eval = F}
# NOTE: Functions take some time to run

# Test on data of length 100 samples -------------------------------------------
# This performs 1,000 iterations of auto_rate on a "default"-type data
set.seed(123)
default100 <- test_lin(reps = 1000, len = 100, type = "default")

# This performs 1,000 iterations of auto_rate on a "corrupted"-type data
set.seed(456)
corrupted100 <- test_lin(reps = 1000, len = 100, type = "corrupted")

# This performs 1,000 iterations of auto_rate on a "segmented"-type data
set.seed(789)
segmented100 <- test_lin(reps = 1000, len = 100, type = "segmented")

# Test on data of length 200 samples -------------------------------------------
# This performs 1,000 iterations of auto_rate on a "default"-type data
set.seed(123)
default200 <- test_lin(reps = 1000, len = 200, type = "default")

# This performs 1,000 iterations of auto_rate on a "corrupted"-type data
set.seed(456)
corrupted200 <- test_lin(reps = 1000, len = 200, type = "corrupted")

# This performs 1,000 iterations of auto_rate on a "segmented"-type data
set.seed(789)
segmented200 <- test_lin(reps = 1000, len = 200, type = "segmented")

# Test on data of length 500 samples -------------------------------------------
# This performs 1,000 iterations of auto_rate on a "default"-type data
set.seed(123)
default500 <- test_lin(reps = 1000, len = 500, type = "default")

# This performs 100 iterations of auto_rate on a "corrupted"-type data
set.seed(456)
corrupted500 <- test_lin(reps = 1000, len = 500, type = "corrupted")

# This performs 100 iterations of auto_rate on a "segmented"-type data
set.seed(789)
segmented500 <- test_lin(reps = 1000, len = 500, type = "segmented")

```

**Comment by nicholascarey** [comment]
Just to add to this, `auto_rate_linear`  which contains `kernel_method` is new. Not sure when added, but it's not in v1.1 code. It seemingly replaces `kde_fit` function, which is now not used anywhere (could be removed).  Both contain this identical line which seems to be where the problem is. 

```
peaks <- which(diff(sign(diff(d$y))) == -2) + 1
```

Not sure why it might be causing a problem in `auto_rate_linear` but not `kde_fit`. Or could be something else entirely....

Tracing back through each, the `roll` originates in `rolling_reg_row`  in new code (`auto_rate_linear`), but in `rolling_reg` in old code (`kde_fit`) so maybe that's something to look at too.

**Comment by nicholascarey** [comment]
Fixed in 7b268c716514d3ab4d1fa5e77c0d129e330e0af1

Error was where zero length data fragments were removed, but same position not removed from the peaks data leading to them having different lengths, so issues when attempting to join them

Code in first comment above no longer produces the error when there are adjacent identical values. No idea why.... 

I still think there's something not quite right here and it may still produce errors....  
Closing for now.
